### PR TITLE
Extension allows multiple same OpTypePointer types

### PR DESCRIPTION
**Pull request by atgoo**
SPV_KHR_variable_pointers allows OpTypePointer to declare multiple
identical types

https://github.com/KhronosGroup/SPIRV-Tools/issues/781

### CODE REVIEW

File: test/val/val_type_unique_test.cpp [hunk 1]
```diff
@@ -238,4 +238,57 @@ OpMemoryModel Physical32 OpenCL
               Not(HasSubstr(GetErrorString(SpvOpTypeVoid))));
 }
 
+TEST_F(ValidateTypeUnique, PointerTypesSameArrayStrideNoExtension) {
```
Comment (dneto0) [inline test/val/val_type_unique_test.cpp]: The test name is misleading to me:  The array strides here are different, but the test name is ....SameArrayStride...

File: source/val/validation_state.cpp [hunk 2]
```diff
@@ -425,6 +426,22 @@ bool ValidationState_t::RegisterUniqueTypeDeclaration(
...
 
+  if (inst.opcode == SpvOpTypePointer &&
+      HasExtension(Extension::kSPV_KHR_variable_pointers)) {
+    const std::vector<Decoration>& decorations = id_decorations(inst.result_id);
+    std::map<SpvDecoration, const std::vector<uint32_t>*> decoration_params;
+    for (const auto& decoration : decorations) {
+      const auto result = decoration_params.emplace(decoration.dec_type(),
+                                                    &decoration.params());
+      (void)result;
+      assert(result.second);
```
Comment (dneto0) [inline source/val/validation_state.cpp]: I don't think you should assert here? There's nothing wrong with redundant decorations, except that it wastes space.
I don't think the storage behind id_decorations has unique'd it.

File: source/val/validation_state.cpp [hunk 3]
```diff
@@ -425,6 +426,22 @@ bool ValidationState_t::RegisterUniqueTypeDeclaration(
     key.insert(key.end(), inst.words + words_begin, inst.words + words_end);
   }
 
+  if (inst.opcode == SpvOpTypePointer &&
```
Comment (dneto0) [inline source/val/validation_state.cpp]: Better to have a comment describing the rule you're enforcing.

File: source/val/validation_state.cpp (hunk 2 above)
Comment (dneto0) [reply source/val/validation_state.cpp]: It's worth saying you're assuming that a single decoration enum can only be set once per target ID.
It think that's true currently, and is likely to be true always.  But I don't recall a SPIR-V validation rule about that. (Perhaps an oversight.)

File: test/val/val_type_unique_test.cpp [hunk 4]
```diff
@@ -238,4 +238,57 @@ OpMemoryModel Physical32 OpenCL
...
+OpDecorate %ptr1 ArrayStride 4
+OpDecorate %ptr2 ArrayStride 8
+%u32 = OpTypeInt 32 0
+%ptr1 = OpTypePointer Input %u32
+%ptr2 = OpTypePointer Input %u32
+)";
+  CompileSuccessfully(str.c_str());
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr(GetErrorString(SpvOpTypePointer)));
```
Comment (dneto0) [inline test/val/val_type_unique_test.cpp]: I'm confused.  The validation succeeds, but generates an error string?

Does this test even pass?

File: source/validate_type_unique.cpp [hunk 5]
```diff
@@ -49,7 +49,7 @@ spv_result_t TypeUniquePass(ValidationState_t& _,
       // return _.diag(SPV_ERROR_INVALID_DATA)
       return _.diag(SPV_SUCCESS)
           << "Duplicate non-aggregate type declarations are not allowed."
-          << " Opcode: " << inst->opcode;
+          << " Opcode: " << spvOpcodeString(SpvOp(inst->opcode));
```
Comment (dneto0) [inline source/validate_type_unique.cpp]: This is an improvement. But in actual practice I've found having the ID generated by this instruction is easier to work with.  It pinpoints the exact instruction at fault.

File: test/val/val_type_unique_test.cpp [hunk 6]
```diff
@@ -238,4 +238,57 @@ OpMemoryModel Physical32 OpenCL
               Not(HasSubstr(GetErrorString(SpvOpTypeVoid))));
 }
 
+TEST_F(ValidateTypeUnique, PointerTypesSameArrayStrideNoExtension) {
+  string str = R"(
+OpCapability Shader
+OpCapability Linkage
+OpMemoryModel Logical GLSL450
+OpDecorate %ptr1 ArrayStride 4
```
Comment (dneto0) [inline test/val/val_type_unique_test.cpp]: Or maybe you just have a typo here?

File: source/val/validation_state.cpp (hunk 2 above)
Comment (atgoo) [reply source/val/validation_state.cpp]: I will assume that decorations are not unique, but only use the last one.

File: source/val/validation_state.cpp (hunk 3 above)
Comment (atgoo) [reply source/val/validation_state.cpp]: Done.

File: source/validate_type_unique.cpp (hunk 5 above)
Comment (atgoo) [reply source/validate_type_unique.cpp]: Done

File: test/val/val_type_unique_test.cpp (hunk 1 above)
Comment (atgoo) [reply test/val/val_type_unique_test.cpp]: CopyPaste

File: test/val/val_type_unique_test.cpp (hunk 4 above)
Comment (atgoo) [reply test/val/val_type_unique_test.cpp]: It logs a warning but returns SPV_SUCCESS.

File: source/val/validation_state.cpp [hunk 7]
```diff
@@ -425,6 +426,25 @@ bool ValidationState_t::RegisterUniqueTypeDeclaration(
...
+      HasExtension(Extension::kSPV_KHR_variable_pointers)) {
+    const std::vector<Decoration>& decorations = id_decorations(inst.result_id);
+    std::map<SpvDecoration, const std::vector<uint32_t>*> decoration_params;
+    for (const auto& decoration : decorations) {
+      decoration_params[decoration.dec_type()] = &decoration.params();
+    }
+    for (const auto& kv : decoration_params) {
+      // Append decoration data to the key, sorted by decoration type.
+      // Use magic number to separate unrelated chunks of data.
+      const uint32_t kMagicSeparator = 0xffff9296;
```
Comment (dneto0) [inline source/val/validation_state.cpp]: Low level comment:  This magic number mechanism is fragile.  Better to push a kv.second.size() then its contents.

(But really we should not care about duplication of decorations.)

File: source/val/validation_state.cpp [hunk 8]
```diff
@@ -425,6 +426,25 @@ bool ValidationState_t::RegisterUniqueTypeDeclaration(
     key.insert(key.end(), inst.words + words_begin, inst.words + words_end);
   }
 
+  // If module has extension SPV_KHR_variable_pointers, then OpTypePointer
+  // is allowed to create identical types so long as their decorations differ.
```
Comment (dneto0) [inline source/val/validation_state.cpp]: Looking again, this is not the rule.
The SPV_KHR_variable_pointers extension says 
> Pointer types are also allowed to have multiple id's for the same opcode and operands, to allow for differing ArrayStride Array Stride decoration values.

> Also, in that paragraph, "non-aggregate types" will then generally be "non-aggregate non-pointer types."

From 2.8 in the SPIR-V spec:
> It is valid to declare multiple aggregate type <id>s having the same opcode and operands. This is to allow multiple instances of aggregate types with the same structure to be decorated differently. (Different decorations are not required; two different aggregate type ids are allowed to have identical declarations and decorations, and will still be two different types.)

That means the validator should allow distinct OpTypePointer with the same operands, and with the same decorations.

Comment (dneto0) [reply source/val/validation_state.cpp]: So I think all this new code boils down to:
```
if (inst.opcode == SpvOpTypePointer && HasExtension....) return true;
```
Even better:  put the check up in the callee, along with the aggregate types.  That is, modify TypeUniquePass to do the pointer-and-has-extension check.

**Comment by atgoo** [comment]
Ready for review.

**Comment by dneto0** [comment]
Hmm.  Except that the commit message was inaccurate.  I'll fix it on the rebase.

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as 725284c2ef8

**Pull request closed without merging**